registry/storage/driver/swift/swift.go

```diff
@@ -298,14 +298,40 @@ func (d *driver) Reader(ctx context.Context, path string, offset int64) (io.Read
 	headers := make(swift.Headers)
 	headers["Range"] = "bytes=" + strconv.FormatInt(offset, 10) + "-"
 
-	file, _, err := d.Conn.ObjectOpen(d.Container, d.swiftPath(path), false, headers)
-	if err == swift.ObjectNotFound {
-		return nil, storagedriver.PathNotFoundError{Path: path}
-	}
-	if swiftErr, ok := err.(*swift.Error); ok && swiftErr.StatusCode == http.StatusRequestedRangeNotSatisfiable {
-		return ioutil.NopCloser(bytes.NewReader(nil)), nil
+	waitingTime := readAfterWriteWait
+	endTime := time.Now().Add(readAfterWriteTimeout)
+
+	for {
+		file, headers, err := d.Conn.ObjectOpen(d.Container, d.swiftPath(path), false, headers)
+		if err != nil {
+			if err == swift.ObjectNotFound {
+				return nil, storagedriver.PathNotFoundError{Path: path}
+			}
+			if swiftErr, ok := err.(*swift.Error); ok && swiftErr.StatusCode == http.StatusRequestedRangeNotSatisfiable {
+				return ioutil.NopCloser(bytes.NewReader(nil)), nil
+			}
+			return file, err
+		}
+
+		//if this is a DLO and it is clear that segments are still missing,
+		//wait until they show up
+		_, isDLO := headers["X-Object-Manifest"]
+		size, err := file.Length()
+		if err != nil {
+			return file, err
+		}
+		if isDLO && size == 0 {
+			if time.Now().Add(waitingTime).After(endTime) {
+				return nil, fmt.Errorf("Timeout expired while waiting for segments of %s to show up", path)
+			}
+			time.Sleep(waitingTime)
+			waitingTime += 2
```

Comment:
The logic in `waitForSegmentsToShowUp` is using `waitingTime *= 2` doubling the wait time each time. Maybe the `*=` was swapped for `+=` by mistake?